modoboa/lib/email_utils.py

```diff
@@ -146,85 +148,124 @@ def _parse(self, msg, level=None):
-                nlevel = level is None and ("%d" % cpt) \
-                    or "%s.%d" % (level, cpt)
-                self._parse(part, nlevel)
-                cpt += 1
-            return
+        content_type = msg.get_content_maintype()
+        is_attachment = "attachment" in msg.get("Content-Disposition", "")
 
-        if level is None:
-            level = "1"
-        try:
-            getattr(self, "_parse_%s" % msg.get_content_maintype())(msg, level)
-        except AttributeError:
-            self._parse_default(msg, level)
+        if content_type == "text" and not is_attachment:
+            self._parse_text(msg, level)
+        elif content_type == "multipart" and not is_attachment:
+            self._parse_multipart(msg, level)
+        else:
+            self._parse_attachment(msg, level)
 
-    def __save_image(self, fname, part):
-        """Save an inline image on the filesystem.
+        if self.contents["plain"]:
+            self._post_process_plain()
+        if self.contents["html"]:
+            self._post_process_html()
 
-        Some HTML messages are using inline images (attached images
-        with links on them inside the body). In order to display them,
-        images are saved on the filesystem and links contained in the
-        message are modified.
+    def _parse_text(self, msg, level=0):
+        content_type = msg.get_content_subtype()
 
-        :param fname: the image associated filename
-        :param part: the email part that contains the image payload
-        """
-        if re.search(r"\.\.", fname):
-            return None
-        path = "/static/tmp/" + fname
-        fp = open(settings.MODOBOA_DIR + path, "wb")
-        fp.write(part.get_payload(decode=True))
-        fp.close()
-        return path
-
-    def map_cid(self, url):
-        """Replace attachment links.
-
-        :param str url: original url
-        :rtype: string
-        :return: internal link
-        """
-        match = re.match(".*cid:(.+)", url)
-        if match is not None:
-            if match.group(1) in self.attached_map:
-                return self.attached_map[match.group(1)]
-        return url
+        if content_type in ["plain", "html"]:
+            encoding = msg.get_content_charset() or "utf-8"
+            mail_text = smart_text(msg.get_payload(decode=True), encoding=encoding, errors="replace")
+            self.contents[content_type] += mail_text
+        else:
+            self._parse_attachment(msg, level)
+
+    def _parse_multipart(self, msg, level=0):
+        level += 1
+        for part in msg.walk():
+            content_type = part.get_content_maintype()
+            if content_type == "text":
+                self._parse_text(part, level=level)
+
+            else:
+                # I'm  a dumb mail parser and treat all non-text parts as attachments
+                self._parse_attachment(part, level=level)
+
+    def _parse_attachment(self, msg, level=0):
```

Comment:
This method is wrongly named since it only deals with images. Talking about that, looks like you don't store attachments anymore (in ``self.attachments``), right?

Reply:
The old code didn't do anything with attachments either it just saved the inline images to disk to be used when the message was displayed.

Reply:
Yes, I checked that yesterday. Attachments were not used since the webmail has its own way to retrieve them. So, I guess we can rename this method to _parse_image or _parse_inline_image?